osmnx/simplification.py

```diff
@@ -451,7 +451,9 @@ def consolidate_intersections(
     dead_ends: bool = False,
     reconnect_edges: bool = True,
     node_attr_aggs: dict[str, Any] | None = None,
+    tolerance_attribute=None,
```

Comment:
Let's simplify this. Rather than introducing a new `tolerance_attribute` function parameter, please change the type hint of the existing `tolerance` parameter to `tolerance: float | dict[int, float] = 10`. This allows the user to either pass a single tolerance value, or a mapping of node IDs -> tolerance values. The argument can be handled accordingly with a simple conditional when it comes time to execute the buffer operation.

Reply:
I thought about this and it was one of the options discussed in https://github.com/gboeing/osmnx/issues/1150 (here: https://github.com/gboeing/osmnx/issues/1150#issuecomment-2062006902). I went for the second option, because I think it's more robust to have the tolerance values attached to the nodes themselves, than keep track of them in a dict. That means you have a separate data structure you have to keep up to date etc.

It also introduces cases like "what if a node id is missing in the dict"? Now it falls back on the default value, otherwise you need to handle that.

I'm open to discussing this further, let me know what you think.

Reply:
Let's keep it as a single `tolerance` parameter that accepts float or dict.

A few responses below.

> it's more robust to have the tolerance values attached to the nodes themselves, than keep track of them in a dict. That means you have a separate data structure you have to keep up to date etc.

The benefit of accepting a tolerance dict is that users *can* still store the tolerance values as node attributes and use that, if they wish, rather than keeping track of a separate data structure. Like this:

```python
ox.consolidate_intersections(G, tolerance=G.nodes(data="tolerance"))
```

So no drawback there. And they can run whatever tolerance estimation algorithm elsewhere then skip the overhead of having to create a special node attribute just to perform the node consolidation, if they wish. It's flexible.

> "what if a node id is missing in the dict"? Now it falls back on the default value, otherwise you need to handle that.

If a node is missing in the dict, it just shouldn't be buffered. The docs can state that explicitly. Users may not want to consolidate intersection in all parts of the network.

The function would work something like:

```python
# ...if the passed tolerance argument is type dict instead of float
gdf_nodes = ox.convert.graph_to_gdfs(G, edges=False)
tolerances = pd.Series(tolerance).reindex(gdf_nodes.index)
gdf_nodes.buffer(tolerances).fillna(gdf_nodes["geometry"])
```

Reply:
Thanks for the detailed response. I agree, this is a cleaner API!

I will try to implement it later today.